source/material_model/melt_boukare.cc

```diff
@@ -0,0 +1,1260 @@
+          reaction_time_step_size = this->get_timestep() / static_cast<double>(number_of_reaction_steps);
+          reaction_fraction = reaction_time_step_size / melting_time_scale;
+        }
+
+
+      const unsigned int Fe_solid_idx = this->introspection().compositional_index_for_name("molar_Fe_in_solid");
+      const unsigned int n_endmembers = endmember_names.size();
+      EndmemberProperties endmembers(n_endmembers);
+
+      for (unsigned int q=0; q<in.temperature.size(); ++q)
+        {
+          std::vector<double> endmember_mole_fractions_per_phase(n_endmembers);
+          std::vector<double> endmember_mole_fractions_in_composite(n_endmembers);
+
+          fill_endmember_properties(in, q, endmembers);
+
+          // We need the compositions of all phases.
+          const double solid_composition = in.composition[q][Fe_solid_idx];
+          double melt_composition = 0.0, melt_molar_fraction = 0.0;
+          double bridgmanite_molar_fraction_in_solid;
+
+          if (this->include_melt_transport())
+            {
+              const unsigned int Fe_melt_idx = this->introspection().compositional_index_for_name("molar_Fe_in_melt");
+              melt_composition = in.composition[q][Fe_melt_idx];
+            }
+
+          convert_composition_to_fraction_of_endmembers(in.temperature[q],
+                                                        solid_composition,
+                                                        melt_composition,
+                                                        endmembers.gibbs_energies,
+                                                        endmember_mole_fractions_per_phase,
+                                                        bridgmanite_molar_fraction_in_solid);
+
+          if (this->include_melt_transport())
+            {
+              const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
+
+              melt_molar_fraction = compute_melt_molar_fraction(in.composition[q][porosity_idx],
+                                                                bridgmanite_molar_fraction_in_solid,
+                                                                endmembers,
+                                                                endmember_mole_fractions_per_phase);
+            }
+
+          const double solid_molar_fraction = 1.0 - melt_molar_fraction;
+
+          // Compute endmember molar fractions in the composite.
+          std::vector<double> phase_mole_fractions_in_composite(n_endmembers);
+          double solid_molar_mass = 0.0;
+          double melt_molar_mass = 0.0;
+          double melt_molar_volume = 0.0;
+          double solid_molar_volume = 0.0;
+
+          for (unsigned int i=0; i<n_endmembers; ++i)
+            {
+              // TODO: fix this using a phase vector (phase == 'bridgmanite')
```

Comment:
I guess this answers part of my comment above. But can you leave a better description than `fix this`? What is `this`?